napari/utils/geometry.py

```diff
@@ -0,0 +1,294 @@
+        Should be a unit vector and have shape (3,).
+
+    Returns
+    -------
+    projected_point : np.ndarray
+        The point that has been projected to the plane.
+        This is always an Nx3 array.
+    """
+    point = np.asarray(point)
+    if point.ndim == 1:
+        point = np.expand_dims(point, axis=0)
+    plane_point = np.asarray(plane_point)
+    # make the plane normals have the same shape as the points
+    plane_normal = np.tile(plane_normal, (point.shape[0], 1))
+
+    # get the vector from point on the plane
+    # to the point to be projected
+    point_vector = point - plane_point
+
+    # find the distance to the plane along the normal direction
+    dist_to_plane = np.multiply(point_vector, plane_normal).sum(axis=1)
+
+    # project the point
+    projected_point = point - (dist_to_plane[:, np.newaxis] * plane_normal)
+
+    return projected_point
+
+
+def rotation_matrix_from_vectors(vec_1, vec_2):
+    """Calculate the rotation matrix that aligns vec1 to vec2.
+
+    Parameters
+    ----------
+    vec_1 : np.ndarray
+        The vector you want to rotate
+    vec_2 : np.ndarray
+        The vector you would like to align to.
+    Returns
+    -------
+    rotation_matrix : np.ndarray
+        The rotation matrix that aligns vec_1 with vec_2.
+        That is rotation_matrix.dot(vec_1) == vec_2
+    """
+    vec_1 = (vec_1 / np.linalg.norm(vec_1)).reshape(3)
+    vec_2 = (vec_2 / np.linalg.norm(vec_2)).reshape(3)
+    cross_prod = np.cross(vec_1, vec_2)
+    dot_prod = np.dot(vec_1, vec_2)
+    if any(cross_prod):  # if not all zeros then
+        s = np.linalg.norm(cross_prod)
+        kmat = np.array(
+            [
+                [0, -cross_prod[2], cross_prod[1]],
+                [cross_prod[2], 0, -cross_prod[0]],
+                [-cross_prod[1], cross_prod[0], 0],
+            ]
+        )
+        rotation_matrix = (
+            np.eye(3) + kmat + kmat.dot(kmat) * ((1 - dot_prod) / (s ** 2))
+        )
+
+    else:
+        if np.allclose(dot_prod, 1):
+            # if the vectors are already aligned, return the identity
+            rotation_matrix = np.eye(3)
+        else:
+            # if the vectors are in opposite direction, rotate 180 degrees
+            rotation_matrix = np.diag([-1, -1, 1])
+    return rotation_matrix
+
+
+def clamp_point_to_bounding_box(point: np.ndarray, bounding_box: np.ndarray):
+    """
+
+    Parameters
+    ----------
+    point : np.ndarray
+        n-dimensional point as an (n,) ndarray
+    bounding_box: np.ndarray
+        n-dimensional bounding box as a (n, 2) ndarray
+
+    Returns
+    -------
+    clamped_point : np.ndarray
+        `point` clamped to the limits of `bounding_box`
+
+    """
+    clamped_point = np.clip(point, bounding_box[:, 0], bounding_box[:, 1] - 1)
+    return clamped_point
+
+
+def face_intercepts_from_bounding_box(bbox: np.ndarray):
+    """TODO: kevin, can you remember what this does?"""
```

Comment:
🤣 ❤️

Reply:
Haha, sorry @alisterburt - I hadn't gotten around to documenting that part of the code yet! This is part of the code to find the intersection of the view ray with the data bounding box. Since our bounding box is axis aligned, we can find the distance to the intersection and thus find the intersection point.

To be a bit more concrete, let's say we have have data with shape` (10, 20, 30)` and we are finding the intersection of the vector `[v_z, v_y, v_x]` origination from point `[p_z, p_y, p_x]` with the plane with normal `[0, 1, 0]` (i.e., the plane described by `y=20`). Given the plane definition, we know that at the intersection point, the y component will be 20, so the displacement in the y direction, `d_y` is

```
d_y = 20 - p_y
```
and the relationship between `d_y` and the distance from the origination point to the intersection point (`d`) is

```
d_y = d * v_y
```
combining those two, we can solve for `d`

```
d = (20 - p_y) / v_y
```
We can then use that distance to find the intersection point

```
intersection_point = [p_z, p_y, p_x + d * [v_z, v_y, v_x]
```

All of that is to say that `face_intercepts_from_bounding_box` is a poorly-named function to find the appropriate values for the bounding box intersection coordinates (20 in the example above and 0 for the other direction) from the bounding box. 😆